src/message/update-firmware/model.tsx

```diff
@@ -0,0 +1,117 @@
+import { Transform, Type } from 'class-transformer';
+import {
+  IsInt,
+  IsNotEmpty,
+  IsOptional,
+  IsString,
+  IsUrl,
+  Length,
+  Min,
+  ValidateNested,
+} from 'class-validator';
+import { Button, Form, Input, Upload } from 'antd';
+import { CustomFormRender } from '../../util/decorators/CustomFormRender';
+import { TransformDate } from '../../util/TransformDate';
+import { UploadOutlined } from '@ant-design/icons';
+
+export enum FirmwareTypeProps {
+  location = 'location',
+  retrieveDateTime = 'retrieveDateTime',
+  installDateTime = 'installDateTime',
+  signingCertificate = 'signingCertificate',
+  signature = 'signature',
+}
+
+export class FirmwareType {
+  @IsNotEmpty()
+  @IsUrl({ require_tld: false })
+  @Length(0, 512)
+  @CustomFormRender(() => {
+    return (
+      <Form.Item
```

Comment:
one thing we can consider in the future is creating a <Url> component and connecting it to the @IsUrl() annotation from class-validator so that we can annotated @IsUrl on a field in the class, and our components can render this field as a file upload

Reply:
This can be done later